lib/push.py

```diff
@@ -129,9 +129,13 @@ def prepare(self):
       self.config = config_bytes
       self.manifest = manifest_bytes
 
-   def push(self):
+   def push(self, prepare_only=False):
       self.prepare()
-      self.upload()
+      if (prepare_only):
+         for (i, (digest, tarball)) in enumerate(self.layers, start=1):
+            ch.INFO("layer %d/%d: %s" % (i, len(self.layers), digest[0:7]))
+      else:
+         self.upload()
```

Comment:
This makes me nervous because it's copy-and-paste code from somewhere in the `self.upload()` call stack. I think it would be better to set a "no uploads" or "dry run" global variable causes a bail-out or no-op somewhere in the upload process, e.g. `Registry_HTTP.request()`, that disables the smallest amount of code possible.

Reply:
Not entirely copy/paste. The block in question exists in `self.upload()` but instead of print information it calls the `ul.layer_from_file()`.

```
144       for (i, (digest, tarball)) in enumerate(self.layers, start=1):
145          ul.layer_from_file(digest, tarball,
146                             "layer %d/%d: " % (i, len(self.layers)))
```

Instead of what I proposed above we could edit this to print the digest info we need, e.g., `layer %d/%d: %s`. However, I don't see a good way to bail out of `Registry_HTTP.request()` without further complicating the call stack, since the function is expected to return a results object. If we have some global here and exit instead of return a response object then subsequent layer info (thinking to the future) will skipped over and lost.

Reply:
Sounds like that's not the right place to bail out then. Where else could you plausibly bail out?